aimon/reprompting_api/pipeline.py

```diff
@@ -0,0 +1,544 @@
+            # Retry LLM call with corrective prompt
+            curr_generated_text = self._call_llm(curr_prompt,self.config.user_model_max_retries, system_prompt, context, user_query)
+            # Re-evaluate the new response
+            curr_payload = self._build_aimon_payload(context, user_query, user_instructions, curr_generated_text, system_prompt)
+            curr_result = self._detect_aimon_response(curr_payload, self.config.feedback_model_max_retries)
+
+            # Extract updated scores and feedback
+            scores, feedback = self.get_response_feedback(curr_result)
+            iteration_num += 1
+            self._record_iteration_output(iteration_outputs, iteration_num, curr_generated_text, curr_result)
+
+        # Final telemetry after loop exit
+        self._emit_iteration_telemetry(
+            iteration_num,
+            pipeline_start,
+            scores,
+            feedback,
+            curr_result,
+            stop_reason or StopReasons.UNKNOWN_ERROR,
+            curr_prompt,
+            curr_generated_text,
+        )
+
+        # Select best response across all iterations
+        best_output, best_failed_count = self._select_best_iteration(iteration_outputs)
+            
+        # Build final response payload
+        response = {"best_response": best_output}
+        if self.config.return_telemetry:
+            response["telemetry"] = self.telemetry.get_all()
+        if self.config.return_aimon_summary:
+            response["summary"] = self._gen_summary(iteration_num, best_failed_count)
+            
+        logger.info("RepromptingPipeline run completed")
+        logger.info(f"Best response selected with {best_failed_count} failed instructions remaining.")
+
+        return response  
+
+    def _build_original_prompt(self) -> Template:
+        """
+        Build a reusable template for combining system_prompt, context, and user_query.
+        This returns a string.Template object so the caller can safely substitute values.
+        
+        Placeholders:
+            - system_prompt
+            - context
+            - user_query
+
+        Returns:
+            Template: A string.Template for building the base LLM prompt.
+        """
+        template_str = (
+            "System:\n${system_prompt}\n\n"
+            "Context:\n${context}\n\n"
+            "User Query:\n${user_query}"
+        )
+        return Template(template_str)
+
+    def _build_aimon_payload(self, context, user_query, user_instructions, generated_text, system_prompt):
+        """
+        Constructs AIMon input payload.
+
+        Args:
+            context (str): Context for the LLM.
+            user_query (str): The user's query.
+            user_instructions (list[str]): Instructions for the model.
+            generated_text (str): The model's generated response.
+
+        Returns:
+            dict: Payload for AIMon evaluation.
+        """
+        if not isinstance(user_instructions, list):
```

Comment:
when the specified user_instructions is empty, the pipeline only uses groundedness and toxicity in the reprompting loop? If yes, this is fine. I would recommend documenting it pipeline.run() function